### ISSUE TITLE
Add an Enumerable.Match method

### ISSUE THREAD

**Issue by simon-curtis** [dotnet/runtime#39443]
## Background and Motivation

Simple proposal to return both matched and unmatched results from a where query in a one-liner.

```c#
var chars = new char[] { 'a', 'b', 'c' };

(matched, unmatched) = chars.Partition(c => c == 'a');
for (var element in unmatched) ...

//or

IMatchResults results = chars.Partition(c => c == 'a');
for (var element in results.Unmatched) ...
```

For readability I would normally do the following; but requires and extra pass
```c#
var chars = new char[] { 'a', 'b', 'c' };
var matched = chars.Where(c => c == 'a');
var unmatched = chars.Where(c => c != 'a');
```

You can achieve this in one pass using a for loop, but goes against the point of Linq and is longer
```c#
var chars = new char[] { 'a', 'b', 'c' };
var matched = new List<char>();
var unmatched = new List<char>();
foreach (var c in chars) {
    if (c == 'a') {
        matched.Add(c);
    } else {
        unmatched.Add(c);
    }
}
```

## Proposed API

The proposed method buffers the results. The returned type is a tuple of `IReadOnlyList`s to better communicate this fact.

```diff
namespace System.Linq
{
    public static class Enumerable
    {
+        public static (IReadOnlyList<TSource> Satisfied, IReadOnlyList<TSource> Falsified) Partition<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate);
    }
}
```

We should _not_ be implementing an IQueryable overload.

**Comment by 333fred** [comment]
This is an API request, moving to the correct repo.

**Comment by simon-curtis** [comment]
@333fred oh sorry, at some point it must have sent me to the runtime repo... I will close and re-open in the correct repo

**Comment by 333fred** [comment]
@simon-curtis sorry if that wasn't clear. I moved it already.

**Comment by canton7** [comment]
Just a note, there was a big discussion about this [on gitter](https://gitter.im/dotnet/csharplang?at=5e6a313202dbe022a180ee67) earlier this year. Turns out it's tricky.

One problem is that Linq wants to be lazy (where possible) and avoid iterating the source collection twice. It's hard to do a partition like this in a lazy way, because it depends on when you decide to enumerate `matched` and `unmatched`: you inevitably end up having to buffer.

Another problem is that `IEnumerator<T>` is `IDisposable` (and things like `File.ReadLines` rely on this to close a resource when enumeration finishes). So we have to make sure that the enumerator returned from `chars.GetEnumerator()` is appropriately disposed when you've finished iterating both `matched` and `unmatched`.

Another is, what happens if you call `matched.GetEnumerator()` or `unmatched.GetEnumerator()` a second time? Do you get another pair of "connected" enumerators?

(You can do something ugly like [this](https://dotnetfiddle.net/IikKxm) to try and overcome those, but it has its own problems.)

**Comment by simon-curtis** [comment]
> @simon-curtis sorry if that wasn't clear. I moved it already. 

Oh I see, thanks! Sorry, still reasonably new to GitHub

**Comment by jnm2** [comment]
Basing it on IEnumerable is dubious as mentioned, but declaring this for concrete collection types like ImmutableArray and ImmutableList is very handy. I call my extension methods `Partition` like Rust's.

**Comment by Clockwork-Muse** [comment]
[`GroupBy`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=netcore-3.1#System_Linq_Enumerable_GroupBy__2_System_Collections_Generic_IEnumerable___0__System_Func___0___1__) can at least get you _close_, as a workaround.  Note this won't be two separate iterators.  Or (ab)using `Join` for a full-lazy iterator (although results won't be grouped by partition).

**Comment by eiriktsarpalis** [comment]
The [F# equivalent](https://msdn.microsoft.com/visualfsharpdocs/conceptual/list.partition%5b%27t%5d-function-%5bfsharp%5d) is also called `partition`.  In C# it might look as follows:
```csharp
public static class Enumerable
{
    public static (List<TSource> Satisfied, List<TSource> Falsified) Partition<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate);
}

**Comment by slavanap** [comment]
What about this approach?

```//Rextester.Program.Main is the entry point for your code. Don't change it.
//Microsoft (R) Visual C# Compiler version 2.9.0.63208 (958f2354)

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace Rextester
{
    public static class MatchExtension {
        public static (IEnumerable<T>, IEnumerable<T>) Match<T>(this IEnumerable<T> c, Func<T, bool> cond) {
            var dict = c.GroupBy(cond).ToDictionary (k => k.Key, v => v);
            return (dict[true], dict[false]);
        }
    }
    public class Program
    {
        public static void Main(string[] args)
        {
            var c = new [] {"a","b","a"};
            var (matched, unmatched) = c.Match(k => k=="a");
            Console.WriteLine(string.Join(",",matched));
            Console.WriteLine(string.Join(",",unmatched));
        }
    }
}```

**Comment by jnm2** [comment]
.ToLookup(...) is better than .GroupBy(...).ToDictionary(...). More efficient is:

```cs
foreach (var value in source)
{
    (predicate(value) ? trueBuilder : falseBuilder).Add(value);
}
```

**Comment by slavanap** [comment]
@jnm2 good point!

To sum up, currently there are 2 solutions for the Match problem:
1) copy (cache) elements into new collections (.ToLookup, 3rd code chunk in @simon-curtis message),
2) run 2 iterators over original IEnumerable (for example, if IEnumerable created from a file) (2rd code chunk in @simon-curtis message)

```C#
using System;
using System.Collections.Generic;
using System.Linq;

public static class MatchExtension {
    public static (IEnumerable<T>, IEnumerable<T>) MatchBuffered<T>(this IEnumerable<T> c, Func<T, bool> cond) {
        var l = c.ToLookup(cond);
        return (l[true], l[false]);
    }
	
    public static (IEnumerable<T>, IEnumerable<T>) Match<T>(this IEnumerable<T> c, Func<T, bool> cond) {
        return (c.Where(cond), c.Where(v => !cond(v)));
    }
}

public class Program
{
    public static void Main(string[] args)
    {
        var c = new [] {"a","b","a"};
        {
            var (matched, unmatched) = c.Match(k => k=="a");
            Console.WriteLine(string.Join(",",matched));
            Console.WriteLine(string.Join(",",unmatched));
        }
        {
            var (matched, unmatched) = c.MatchBuffered(k => k=="a");
            Console.WriteLine(string.Join(",",matched));
            Console.WriteLine(string.Join(",",unmatched));
        }
    }
}
```

The non-buffered approach can be extended for IQueryable, but would require a bit code for original Expression negation.

**Comment by simon-curtis** [comment]
I had this thought from a project I’m doing in blazor with EFCore, where I was separating entities’ UI elements by its archived property. 

‘’’c#
foreach (var entity in matched) ...
foreach (var entity in unmatched) ...
‘’’

I wouldn’t want to go back to the database on every render so would definitely want some form of buffering, in this instance I’m using DbContext. But I could see how in other cases you wouldn’t want buffering.

I did end up using GroupBy().ToDictionary(), just not as a static method so I will do that on Monday to make my life easier than doing it every time, but looked messy and that’s why I’m suggesting a one-liner

**Comment by jnm2** [comment]
How would the non-buffered approach work when the matched or unmatched enumerables are enumerated more than once?

Assume that `source` returns a different sequence on each enumeration, which enumerables are explicitly allowed to do:

```cs
var (matched, unmatched) = source.Partition(cond);

var a = matched.Select(...).ToList(); // Fully enumerate matched enumerable once
var b = unmatched.Select(...).ToList(); // Fully enumerate unmatched enumerable once

var c = matched.Select(...).ToList();  // Fully enumerate matched enumerable a second time
var d = unmatched.Select(...).ToList();  // Fully enumerate matched enumerable a second time
```

You would assume that elements in `a` and `b` are based on the same source sequence and are correlated by index. But what about `c` and `d`?

**Comment by slavanap** [comment]
For non-buffered approach I've personally assumed that the source sequence is stable, i.e. does not change over time. For files & databases this is easily achievable. For random generators - of course, not.

**Comment by jnm2** [comment]
For databases, enumerating `source` a second time means querying the database a second time. I might be wrong, but I think the way to state through strong typing that a sequence is stable is `IReadOnlyCollection<T>`.

**Comment by slavanap** [comment]
@jnm2 , yes. Enumerating database twice will query database twice. You can lock database state with a transactions mechanism. But for databases one would need to implement approach with IQueryable to stream data effectively, because conversion to IEnumerable will trigger data stream of full results set over a network.

**Comment by jnm2** [comment]
If the method accepts `IEnumerable<T>`, people will certainly pass enumerables that return different sequences on each enumeration. This isn't far-fetched from the code I've seen. If the Partition method doesn't buffer, what should the returned enumerables do if they are enumerated more than once?

**Comment by slavanap** [comment]
Usually it depends on the implementation. IEnumerable is allowed to be enumerated more than once.

**Comment by jnm2** [comment]
@slavanap My worry is, what should that implementation be? I'm not sure how to avoid buffering and be safe. And you can't get away from all buffering. If the first half doesn't match and the second half does, and you fully enumerate the matched enumerable first, you have to buffer half the data.

**Comment by slavanap** [comment]
@jnm2 you dispose data while enumerating (and with IQueryable you query each half separately without loading any data from another half).

If source IEnumerable sources data from file or network, then your IEnumerable size is not constrained by RAM capacity.

While buffered IEnumerable is simpler it has its limitations, like RAM constraints.

**Comment by jnm2** [comment]
I am not comfortable with this concept of querying each half separately by default. Especially not after the single-query discussion in EF Core (https://github.com/dotnet/efcore/issues/18022).

File and network also seem to have concurrency issues close enough to a database unless you have exclusive control. And still, how could the Partition method avoid buffering the source enumerable when it is backed by a file? (If the first half doesn't match and the second half does and you fully enumerate the matched enumerable first.)

**Comment by Clockwork-Muse** [comment]
@jnm - Well, a custom one for a file could potentially have two file pointers, which would allow you to avoid buffering.

That said, I would regard any `IQueryable` that performed two queries for something like this to be unsafe.  I would far rather either branch in the `foreach` (or do something like one of the workarounds to generate it), or realize the list as part of the partition.  If the list would be too big to hold in memory, you'd have to lock for too long for it to be safe - and at that point I might be trying to do any sort of update DB side anyways.

**Comment by eiriktsarpalis** [comment]
My personal preference would be for a buffered approach, in order to avoid double enumeration. That being said, I am concerned that this violates LINQ convention (all methods returning IEnumerable are "lazy"). I also don't think it would make sense to expose an `IQueryable` variant.

So perhaps the best course of action is to close this issue. It can be worked around trivially using `GroupBy` or `ToLookup`.

**Comment by jnm2** [comment]
@eiriktsarpalis Is what Match would be doing any more of a violation than what GroupBy does when it returns `IGrouping<T>` (`IEnumerable<T>`) instances?

**Comment by eiriktsarpalis** [comment]
Well the key difference is that no groupings will be calculated unless you create an enumerator. There is no clear way to achieve the same semantics when you return a pair of enumerables.

**Comment by jnm2** [comment]
@eiriktsarpalis I see. It's like enumerating the first two groupings from a GroupBy, and the latter is not lazy. The syntax sugar is worth it to me. Could we see `Match` in the same category of `Enumerable` methods as `ToLookup` and `ToList` and just be fine with the fact that it's not lazy?

**Comment by eiriktsarpalis** [comment]
> Could we see Match in the same category of Enumerable methods as ToLookup and ToList and just be fine with the fact that it's not lazy?

We could, but I think the return type should communicate this fact by returning something like `IReadOnlyCollection<T>` instead of `IEnumerable<T>`.

**Comment by jnm2** [comment]
@eiriktsarpalis I'd be a fan of that. Or `ImmutableArray<T>` to be honest.

**Comment by eiriktsarpalis** [comment]
> Or ImmutableArray<T> to be honest.

Hmm, presumably to eliminate virtual calls? `GroupBy` and `ToLookup` return interfaces. To my knowledge the only LINQ APIs returning concrete collection types are the `To[CollectionType]()` methods so we might want to stick to that convention. Wouldn't mind using `ImmutableArray<T>` as the backing implementation but the return type should probably be an interface. We might want to consider `IReadOnlyList<T>`.

**Comment by jnm2** [comment]
Just my favorite inter-method collection type. I'm not strongly tied to it.

I think of `IReadOnlyList<T>` as better than `IReadOnlyCollection<T>` if the semantics are such that the elements are ordered in some way rather than being specifically unordered.

**Comment by eiriktsarpalis** [comment]
Ok so something like this then:

```csharp
public static class Enumerable
{
    public static (IReadOnlyList<TSource> Satisfied, IReadOnlyList<TSource> Falsified) Partition<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
    {
        ImmutableArray<TSource>.Builder satisfied = ImmutableArray.CreateBuilder<TSource>();
        ImmutableArray<TSource>.Builder falsified = ImmutableArray.CreateBuilder<TSource>();

        foreach (TSource value in source)
        {
            (predicate(value) ? satisfied : falsified).Add(value);
        }

        return (satisfied.ToImmutable(), falsified.ToImmutable());
    }
}
```
If folks agree with this signature I will update the OP and mark as ready for review.

**Comment by jnm2** [comment]
@eiriktsarpalis Signature looks good to me. I wouldn't implement using `ImmutableArray<TSource>.Builder` unless the signature required ImmutableArray to be returned, probably `List<TSource>` instead, but there's probably prior art within LINQ that decides what gets used for the implementation anyway.

**Comment by eiriktsarpalis** [comment]
@simon-curtis I have updated your original post to include the updated proposal.

**Comment by MgSam** [comment]
I'm not really convinced this method is necessary. It's just a special case of GroupBy where there are only two groups. Is making a special method for that really worth it?

**Comment by MgSam** [comment]
Another concern here is that the word "Partition" is basically a synonym for "Group". So it can become confusing to users why/how the two methods are different.

**Comment by jnm2** [comment]
@MgSam People learn what it effectively means and get used to terminology like this in language keywords and other places in the BCL, too.

**Comment by jnm2** [comment]
> I'm not really convinced this method is necessary. It's just a special case of GroupBy where there are only two groups. Is making a special method for that really worth it?

Yes, because it's too much ceremony to pull out the true and false groups (if any) and put them in variables to work on.

**Comment by KieranDevvs** [comment]
> I'm not really convinced this method is necessary. It's just a special case of GroupBy where there are only two groups. Is making a special method for that really worth it?

I would have to agree that this seems like something that isn't common enough, nor is the use case consistent enough to warrant a given implementation. For example, if this were to be implemented, I would expect the properties `Satisfied` and `Unsatisfied` to be deferred, which causes performance concerns if you want to enumerate both collections. However, having the result buffered introduces unexpected behaviour in the sense that I would expect an enumerable collection to be able to change (which will have side effects in some peoples use cases).

To summarise... I personally think this should not be part of the BCL and should be left to the user to implement, as implementations vary significantly while providing no real benefit other than syntactic short-handing.

**Comment by jnm2** [comment]
My understanding is that there are not so many ways this could be done. Returning lazy enumerables has issues as discussed above, and you lose the analogy to GroupBy groupings which are also not lazy enumerables as mentioned above. The syntactic simplicity *is* the request, because this is not fun:

```cs
var usersByActivityStatus = usersEnumerable.GroupBy(user => user.IsActive);
var activeUsers = (usersByActivityStatus.SingleOrDefault(group => group.Key) ?? Enumerable.Empty<User>()).ToList();
var inactiveUsers = (usersByActivityStatus.SingleOrDefault(group => !group.Key) ?? Enumerable.Empty<User>()).ToList();

// Loop multiple times over, or index into, activeUsers and inactiveUsers
```

Nor is this:

```cs
var usersByActivityStatus = usersEnumerable.ToLookup(user => user.IsActive);
var activeUsers = usersByActivityStatus[true].ToList();
var inactiveUsers = usersByActivityStatus[false].ToList();

// Loop multiple times over, or index into, activeUsers and inactiveUsers
```

Nor is this:

```cs
var activeUsers = usersEnumerable.Where(group => group.Key).ToList();
var inactiveUsers = usersEnumerable.Where(group => !group.Key).ToList();

// Loop multiple times over, or index into, activeUsers and inactiveUsers
```

This is not rare or contrived. We run into this organically. I think others could benefit from this syntax:

```cs
var (activeUsers, inactiveUsers) = usersEnumerable.Partition(group => group.Key);

// Loop multiple times over, or index into, activeUsers and inactiveUsers
```

**Comment by KieranDevvs** [comment]
@jnm2 

```C#
var usersByActivityStatus = usersEnumerable.GroupBy(user => user.IsActive);
```

Is equivalent to 

```C#
var (activeUsers, inactiveUsers) = usersEnumerable.Partition(group => group.Key);
```

The only difference being is the return type structure. One being two collections and one being a single key collection.
I dont see how:

```
var usersByActivityStatus = usersEnumerable.GroupBy(user => user.IsActive);
var activeUsers = usersByActivityStatus[true].ToList();
var inactiveUsers = usersByActivityStatus[false].ToList();
```

Is any different from

```
var (activeUsers, inactiveUsers) = usersEnumerable.Partition(group => group.Key);
var activeUsersCollection = activeUsers.ToList();
var inactiveUsersCollection = inactiveUsers.ToList();
```

Other than the fact you have to do a hash table lookup between two values of true and false (which is trivial).

**Comment by jnm2** [comment]
It's different because you don't need the ToList calls that you added, per the proposal above. It's an actual one-liner. The return structure is the entire point of the proposal.

**Comment by KieranDevvs** [comment]
> It's different because you don't need the ToList calls that you added, per the proposal above that says there should be no deferral just like how IGroupings returned by GroupBy don't themselves have deferral. It's an actual one-liner. The return structure is the entire point of the proposal.

So you're saving a `.ToList()`...
Create an extension method in your application, this really isn't warranted.

**Comment by jnm2** [comment]
Saving two ToLists and performing better than GroupBy.

**Comment by KieranDevvs** [comment]
> Saving two ToLists and performing better than GroupBy.

How does it perform better?
Unless I'm mistaken, the LINQ `GroupBy` results in one iteration.
![image](https://user-images.githubusercontent.com/21192520/105899469-9fb43f80-6012-11eb-8ddd-34920f6b0a70.png)

**Comment by KieranDevvs** [comment]
> Saving two ToLists and performing better than GroupBy/ToLookup while enumerating the original enumerable just once.

Hmm ok, after consideration, I see the performance impact. Maybe there is a benefit, I'm just still not convinced that this isn't a niche problem. I guess if there are large quantities of people, commonly using this Boolean grouping pattern, then I'd be on board.

**Comment by simon-curtis** [comment]
Would I be right in saying this would be the implementation?

```C#
public static class Extensions
{
    public static (IReadOnlyList<TSource> Satisfied, IReadOnlyList<TSource> Falsified) Partition<TSource>(
        this IEnumerable<TSource> source, Func<TSource, bool> predicate)
    {
        var results = source.ToLookup(predicate.Invoke);
        return (results[true].ToList(), results[false].ToList());
    }
}
```

@KieranDevvs I know it's trivial but I enjoy the deconstruction ability over two lines - although as you say, it still does the same thing.

Also, I love records but don't get to use them enough as most of our codebases are "legacy" 👎

```C#
public record PartitionResult<TSource>(IReadOnlyList<TSource> Satisfied, IReadOnlyList<TSource> Falsified)
```

**Comment by eiriktsarpalis** [comment]
> Would I be right in saying this would be the implementation?

I think we could avoid all the lookups if we followed an approach as described here: https://github.com/dotnet/runtime/issues/39443#issuecomment-738084914

**Comment by terrajobst** [comment]
* Let's return the raw `List<T>` because it's more useful to callers (and we likely can't change the instance anyway)
* Let's rename the tuple fields to `Matched` and `Unmatched` to make easier to understand

```C#
namespace System.Linq
{
    public partial static class Enumerable
    {
        public static (List<TSource> Matched, List<TSource> Unmatched) Partition<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate);
    }
}
```

**Comment by MgSam** [comment]
I strongly believe you should return `IReadOnlyList<T>` and not `List<T>`. You are constraining your implementation in now until forever if you return the concrete `List<T>` type. The method signature should not be dictated by the implementation whenever possible.

**Comment by eiriktsarpalis** [comment]
> You are constraining your implementation in now until forever if you return the concrete List<T> type. The method signature should not be dictated by the implementation whenever possible.

I don't see how this implementation would ever use anything other than `List<T>`, so we opted for the simple solution. There are benefits to returning a concrete type and there is certainly precedent in LINQ, e.g. the `ToDictionary()` method.

**Comment by MgSam** [comment]
>I don't see how this implementation would ever use anything other than List<T>...

We never know now what we'll know in the future. That's the entire point of defensive coding. The framework is littered with stuff that is now considered obsolete or bad practice that seemed like a great idea at the time.

> ...there is certainly precedent in LINQ, e.g. the ToDictionary() method.

`ToDictionary()` and `ToList()` are totally different. They are for the express purpose of getting those concrete types back. 

No other LINQ methods return concrete types. Including GroupBy, where you are actually constructing a completely new collection but hiding it behind an interface.

>  There are benefits to returning a concrete type...

Like what? The only benefit that I can think of is that the user will mutate it, which is not something the BCL should be encouraging. 

Returning `IReadOnlyList<T>` is a much safer and better practice and should be the example the BCL sets. It is ordered, it has a count; the only thing it cannot do is mutate. 

Returning `List<T>` from **any** public method is a code smell and IMO it should absolutely be not baked into a BCL LINQ method.

**Comment by eiriktsarpalis** [comment]
>> I don't see how this implementation would ever use anything other than List...
>
> We never know now what we'll know in the future. That's the entire point of defensive coding. The framework is littered with stuff that is now considered obsolete or bad practice that seemed like a great idea at the time.

While I understand the point you are trying to make in principle, I don't agree that it applies to this particular case.

> ToDictionary() and ToList() are totally different. They are for the express purpose of getting those concrete types back.
>
> No other LINQ methods return concrete types. Including GroupBy, where you are actually constructing a completely new collection but hiding it behind an interface.

Note that the new method has more similarities with `ToDictionary` and `ToList` than with `GroupBy: it eagerly consumes the source enumerable and returns a materialized collection. We actually debated following the `ToType` naming convention but just couldn't find an appropriate name. This is also why the method doesn't come with an accompanying Queryable overload.

> Returning List<T> from any public method is a code smell and IMO it should absolutely be not baked into a BCL LINQ method.

Again, I understand the point you are making in principle, but I don't see how that translates to the particular method. What particular problems would you expect this cause for users?

**Comment by jnm2** [comment]
@MgSam This was discussed in the video at https://youtu.be/7bRCbwE9CYE?t=2772. Someone mentions:

> We do have a principle which is basically: parameters should accept the weakest types they need and return types should be the strongest type you have.

**Comment by MgSam** [comment]
I think a better analog rather than `ToDictionary` or `ToList` is `Regex.Matches`. It's doing conceptually the same thing as this Partition method- except it works on a Enumerable<char> (aka, a string) instead. And it returns `MatchCollection`, which is a **read-only** collection of the different partitions.

> We do have a principle which is basically: parameters should accept the weakest types they need and return types should be the strongest type you have.

I totally agree with this principle- which is why I'm proposing `IReadOnlyList<T>` and not `IEnumerable<T>`. Having an ordering and a count are both super useful properties which you lose if you return `IEnumerable`.

> What particular problems would you expect this cause for users?

There are multiple problems:
- Encouraging the bad practice of mutating a collection that wasn't created by you and doesn't belong to you. 
- Baking an inconsistency into the heart of LINQ. Partition is a specialized case of GroupBy, but while GroupBy returns an `IGrouping`, Partition will return a `List<T>`?
- Using a concrete type when an interface would be provide future-proofing. (Thus ensuring that if Eirik Jr in 20 years wants to change the implementation of Partition so that it instead returns some new magical type that uses a quantum cache, he can't do so). 

The interface vs non-interface is the point I still really don't understand and which you haven't answered. Even if we all agreed it was a great idea to return a mutable list as the result of Partition- why wouldn't we use `IList<T>` rather than `List<T>`? At the end of the day the vast majority of the users won't care if the implementation of the result is a `List<T>` or a `PartitionCollection` or an array or an `ImmutableList<T>` under the hood.

**Comment by jnm2** [comment]
@MgSam 

> Encouraging the bad practice of mutating a collection that wasn't created by you and doesn't belong to you.

But since they explicitly wanted the collections created by this API to belong to you, this is a good reason to return a concrete collection type.

**Comment by jnm2** [comment]
@MgSam 

> Encouraging the bad practice of mutating a collection that wasn't created by you and doesn't belong to you.

But since they explicitly wanted the semantics of this API to be that the collections being created do belong to you, this is a good reason to return a concrete collection type.

> Baking an inconsistency into the heart of LINQ. Partition is a specialized case of GroupBy, but while GroupBy returns an `IGrouping`, Partition will return a `List<T>`?

As Eirik said above, this is a matter of perspective. This is intended to diverge from GroupBy because it's intended to serve a slightly different purpose. This usage pattern is not perfectly served by GroupBy, thus the difference and the reason for this API request.

> Using a concrete type when an interface would be provide future-proofing. (Thus ensuring that if Eirik Jr in 20 years wants to change the implementation of Partition so that it instead returns some new magical type that uses a quantum cache, he can't do so).

This was addressed in the video. It's really an interesting listen.

**Comment by MgSam** [comment]
@jnm2 I did listen, and if anything the video proves why this issue deserves more thought. They barely spent about two minutes discussing it where two of the participants opinions are "I don't really care" and the third basically says "let's return List<T> because that's what we're using internally". They do discuss the immutable vs mutable question but don't present a very compelling argument IMO. _The question of why `List<T>` vs `IList<T>` isn't even addressed._  Certainly no one brings up the obvious parallel to Regex.Matches. 

Public methods should not be exposing implementation details. .NET should not change framework patterns arbitrarily because the developers reviewing the code can't be bothered to make an opinion about it or consider it in the wider context of the rest of the framework. 

At the end of the day, it's a minor issue either way but I want to see .NET be the best platform it can be. And I certainly don't want to see LINQ proper become an everything-and-the-kitchen-sink mess of methods with inconsistent usage and documentation like Rx is.

**Comment by eiriktsarpalis** [comment]
> The question of why List<T> vs IList<T> isn't even addressed.

Simply put, this is an elementary helper method for which `List<T>` offers the canonical, optimal solution. There are no "clever" optimizations that we would ever consider including, quantum or otherwise.

> Public methods should not be exposing implementation details.

I think you might be applying a liberal definition of "implementation detail". `List<T>` is a public type that is fairly well understood. Perhaps using `ValueTuple` in the return type is leaking implementation detail too? What about `Enumerable.Count()`  only returning 32-bit integers?

>  .NET should not change framework patterns arbitrarily because the developers reviewing the code can't be bothered to make an opinion about it or consider it in the wider context of the rest of the framework.

What "framework patterns" are you referring to? Here's a quote from the [framework design guidelines](https://github.com/dotnet/runtime/blob/master/docs/coding-guidelines/framework-design-guidelines-digest.md#general-design-guidelines), also mentioned during the meeting:

>  *DO* use the most derived type for return values and the least derived type for input parameters. For example take IEnumerable as an input parameter but return Collection<string> as the return type. Provide a clear API entry point for every scenario. Every feature area should have preferably one, but sometimes more, types that are the starting points for exploring given technology. We call such types Aggregate Components. Implementation of large majority of scenarios in given technology area should start with one of the Aggregate Components.

Ultimately, would you be able to propose a potential future concern, particular to this method, that might justify abstracting the return type? Abstraction comes with tradeoffs, so we won't be adopting it for the sake of an ambiguous sense of futureproofing.

**Comment by canton7** [comment]
I just want to make a small point as an observer -- This is Microsoft's house, and they've kindly invited us in to see how they do things. Remember that they are the owners and this is their job: we can make suggestions and give our personal opinions, but we do not have the right to tell them what to do.

I've seen first-hand the annoyance that some stranger telling you how to do your job causes, and I'm keen that we don't risk ruining the great privilege we have here :)

(I don't want to debate this message, just dropping past on my way though)

**Comment by danmoseley** [comment]
@MgSam 

> because the developers reviewing the code can't be bothered to make an opinion about it 

We welcome your interaction and input here. At the same time, please assume good intent. This is called out in our [code of conduct](https://dotnetfoundation.org/about/code-of-conduct).

**Comment by terrajobst** [comment]
@MgSam 

I just want to elaborate on a few points here:

1. The reason we return a concrete type is because we practically can't change the implementation unless the concrete type implementing the interface is fully internal. The reason being that people very often (accidentally or intentionally) downcast the result from, say, `IList<T>` to `List<T>`, which would break if we suddenly return a different instance. "But that's theoretical". Sadly it's not, we learned that the hard way when we changed `Enumerable.Empty<T>()` from returning an empty `T[]` to an internal type that implements `IEnumerable<T>`, so we undid that change and kept returning `T[]`. That's why @bartonjs and me believe that if the framework returns a public type, we're better of using the concrete type.

2. "The team only spent two minutes on it". That is fair and I'm not gonna claim that the API review team has perfect visibility on all the issues presented to us, including the design trade offs. However, that doesn't mean that nobody does. The domain expert for Linq is @eiriktsarpalis and he presented the issue to us and answered our question. His expertise, and the time he has spent on this thread alone, is far more than the two minutes you cite.

3. We value different view points. As should be obvious to anyone watching our API reviews we ourselves rarely have a uniform view point on issues. In fact, whenever we do I'm slightly nervous that we're missing something because virtually no design is perfect. If you think it is, it usually means you're not aware of some shortcoming, not that it's free of them. I think your concerns and criticisms have merit. But as @danmoseley pointed out tone matters. Rest assured we all get frustrated and concerned when other people don't understand or don't care for our arguments. But snapping and lashing out at people isn't an option because it would meant that 90% of our discussions would become toxic wastelands as consensus never exists at the beginning, it is created through discourse and exchange of criticism.

**Comment by terrajobst** [comment]
@canton7 

> I just want to make a small point as an observer -- This is Microsoft's house, and they've kindly invited us in to see how they do things. Remember that they are the owners and this is their job: we can make suggestions and give our personal opinions, but we do not have the right to tell them what to do.
>
> I've seen first-hand the annoyance that some stranger telling you how to do your job causes, and I'm keen that we don't risk ruining the great privilege we have here :)

Thanks for the kind words. I just want to point that this is a two way street for us: we equally feel very privileged of being lucky that so many contributors take their time and help us make .NET better. That is a big reason why I care about the conduct in our discussions: it's not just how people talk to us, it's how we all talk to each other. I don't want to lose contributors or potential contributors because they see a toxic thread and decide that this isn't the community they want to be a part of.

**Comment by stephentoub** [comment]
> Sadly it's not, we learned that the hard way when we changed Enumerable.Empty<T>() from returning an empty T[] to an internal type that implements IEnumerable<T>, so we undid that change and kept returning T[]

@terrajobst, FYI, we've been returning something other than `T[]` for several releases now.

**Comment by terrajobst** [comment]
> (@terrajobst, just FYI, we've actually been returning something other than `T[]` for several releases now, since .NET Core 3.0 if memory serves.)

Don't ruin my brilliant tale with facts 😄

**Comment by danmoseley** [comment]
`IDictionary GetEnvironmentVariables()` is a better example. It was changed to return a `Dictionary<string, string>` but that was too breaking so it had to go back to a `Hashtable`. This actually broke me on a previous product -- it's so tempting to downcast that `IDictionary`, because you know what it is, right? That isn't to say we never make changes like that, particularly in .NET Core.

**Comment by MgSam** [comment]
Thanks guys for the explanation. I agree there definitely can be unintended consequences even with an interface as the return type if you ever needed to change that, especially at the BCL level. That said, someone's reliance on internal implementation details is a much "easier" breaking change to swallow than changing a concrete type.

**Comment by eiriktsarpalis** [comment]
Per https://github.com/dotnet/runtime/pull/49819#pullrequestreview-615393581 we have reached the conclusion that the proposed method, while useful for some applications, does not meet the bar for inclusion in System.Linq. I'm therefore going to close this issue.
